File: trading_dummy/lib/models/agent_states.dart
// Agent States for Trading Graph
// Mirrors the Python agent_states.py structure

class AgentState {
  String companyOfInterest;
  String tradeDate;
  
  // Analyst Reports
  String marketReport;
  String sentimentReport;
  String newsReport;
  String fundamentalsReport;
  
  // Investment Flow
  String investmentPlan;
  String traderInvestmentPlan;
  String finalTradeDecision;
  
  // Debate States
  InvestDebateState investmentDebateState;
  RiskDebateState riskDebateState;

  AgentState({
    required this.companyOfInterest,
    required this.tradeDate,
    this.marketReport = '',
    this.sentimentReport = '',
    this.newsReport = '',
    this.fundamentalsReport = '',
    this.investmentPlan = '',
    this.traderInvestmentPlan = '',
    this.finalTradeDecision = '',
    InvestDebateState? investmentDebateState,
    RiskDebateState? riskDebateState,
  }) : investmentDebateState = investmentDebateState ?? InvestDebateState(),
       riskDebateState = riskDebateState ?? RiskDebateState();

  AgentState copyWith({
    String? companyOfInterest,
    String? tradeDate,
    String? marketReport,
    String? sentimentReport,
    String? newsReport,
    String? fundamentalsReport,
    String? investmentPlan,
    String? traderInvestmentPlan,
    String? finalTradeDecision,
    InvestDebateState? investmentDebateState,
    RiskDebateState? riskDebateState,
  }) {
    return AgentState(
      companyOfInterest: companyOfInterest ?? this.companyOfInterest,
      tradeDate: tradeDate ?? this.tradeDate,
      marketReport: marketReport ?? this.marketReport,
      sentimentReport: sentimentReport ?? this.sentimentReport,
      newsReport: newsReport ?? this.newsReport,
      fundamentalsReport: fundamentalsReport ?? this.fundamentalsReport,
      investmentPlan: investmentPlan ?? this.investmentPlan,
      traderInvestmentPlan: traderInvestmentPlan ?? this.traderInvestmentPlan,
      finalTradeDecision: finalTradeDecision ?? this.finalTradeDecision,
      investmentDebateState: investmentDebateState ?? this.investmentDebateState,
      riskDebateState: riskDebateState ?? this.riskDebateState,
    );
  }
}

class InvestDebateState {
  String bullHistory;
  String bearHistory;
  String currentResponse;
  String judgeDecision;
  int count;

  InvestDebateState({
    this.bullHistory = '',
    this.bearHistory = '',
    this.currentResponse = '',
    this.judgeDecision = '',
    this.count = 0,
  });

  InvestDebateState copyWith({
    String? bullHistory,
    String? bearHistory,
    String? currentResponse,
    String? judgeDecision,
    int? count,
  }) {
    return InvestDebateState(
      bullHistory: bullHistory ?? this.bullHistory,
      bearHistory: bearHistory ?? this.bearHistory,
      currentResponse: currentResponse ?? this.currentResponse,
      judgeDecision: judgeDecision ?? this.judgeDecision,
      count: count ?? this.count,
    );
  }
}

class RiskDebateState {
  String riskyHistory;
  String safeHistory;
  String neutralHistory;
  String currentRiskyResponse;
  String currentSafeResponse;
  String currentNeutralResponse;
  String judgeDecision;
  int count;

  RiskDebateState({
    this.riskyHistory = '',
    this.safeHistory = '',
    this.neutralHistory = '',
    this.currentRiskyResponse = '',
    this.currentSafeResponse = '',
    this.currentNeutralResponse = '',
    this.judgeDecision = '',
    this.count = 0,
  });

  RiskDebateState copyWith({
    String? riskyHistory,
    String? safeHistory,
    String? neutralHistory,
    String? currentRiskyResponse,
    String? currentSafeResponse,
    String? currentNeutralResponse,
    String? judgeDecision,
    int? count,
  }) {
    return RiskDebateState(
      riskyHistory: riskyHistory ?? this.riskyHistory,
      safeHistory: safeHistory ?? this.safeHistory,
      neutralHistory: neutralHistory ?? this.neutralHistory,
      currentRiskyResponse: currentRiskyResponse ?? this.currentRiskyResponse,
      currentSafeResponse: currentSafeResponse ?? this.currentSafeResponse,
      currentNeutralResponse: currentNeutralResponse ?? this.currentNeutralResponse,
      judgeDecision: judgeDecision ?? this.judgeDecision,
      count: count ?? this.count,
    );
  }
} 

File: trading-graph-server/src/agent/utils/agent_states.py
from typing import Annotated, Sequence
from datetime import date, timedelta, datetime
from typing_extensions import TypedDict, Optional
from langchain_openai import ChatOpenAI
from langgraph.graph import END, StateGraph, START
from langchain_core.messages import BaseMessage
from langgraph.graph.message import add_messages

# TASK 5.2: Import optimized state management
from .state_optimizer import (
    create_optimized_message_reducer,
    create_optimized_report_reducer, 
    create_optimized_debate_reducer,
    get_state_manager
)

# TASK 5.2: Memory-optimized reducers (40% memory reduction target)
optimized_message_reducer = create_optimized_message_reducer()
optimized_report_reducer = create_optimized_report_reducer()
optimized_debate_reducer = create_optimized_debate_reducer()

# Simple reducer that always takes the latest value
def update_value(left, right):
    return right if right is not None else left

# TASK 5.2: Legacy reducers kept for backward compatibility but optimized
def merge_debate_state(left, right):
    """TASK 5.2: Optimized debate state merger with memory efficiency."""
    return optimized_debate_reducer(left, right)

def merge_risk_debate_state(left, right):
    """TASK 5.2: Optimized risk debate state merger with memory efficiency."""
    return optimized_debate_reducer(left, right)


# Researcher team state
class InvestDebateState(TypedDict):
    bull_history: Annotated[str, update_value]
    bear_history: Annotated[str, update_value]
    history: Annotated[str, update_value]
    current_response: Annotated[str, update_value]
    judge_decision: Annotated[str, update_value]
    count: Annotated[int, update_value]


# Risk management team state
class RiskDebateState(TypedDict):
    risky_history: Annotated[str, update_value]
    safe_history: Annotated[str, update_value]
    neutral_history: Annotated[str, update_value]
    history: Annotated[str, update_value]
    latest_speaker: Annotated[str, update_value]
    current_risky_response: Annotated[str, update_value]
    current_safe_response: Annotated[str, update_value]
    current_neutral_response: Annotated[str, update_value]
    judge_decision: Annotated[str, update_value]
    count: Annotated[int, update_value]


class AgentState(TypedDict):
    """
    TASK 5.2: Memory-optimized multi-agent system state with atomic updates
    Target: 40% memory reduction through optimized reducers and state validation
    """
    
    # Basic information
    company_of_interest: Annotated[str, update_value]
    trade_date: Annotated[str, update_value]
    
    # TASK 5.2: Optimized analyst message channels with memory efficiency
    market_messages: Annotated[Sequence[BaseMessage], optimized_message_reducer]
    social_messages: Annotated[Sequence[BaseMessage], optimized_message_reducer]
    news_messages: Annotated[Sequence[BaseMessage], optimized_message_reducer]
    fundamentals_messages: Annotated[Sequence[BaseMessage], optimized_message_reducer]
    
    # TASK 5.2: Optimized reports with conflict resolution and validation
    market_report: Annotated[Optional[str], optimized_report_reducer]
    sentiment_report: Annotated[Optional[str], optimized_report_reducer]
    news_report: Annotated[Optional[str], optimized_report_reducer]
    fundamentals_report: Annotated[Optional[str], optimized_report_reducer]
    
    # TASK 5.2: Optimized debate states with atomic updates and memory efficiency
    investment_debate_state: Annotated[Optional[InvestDebateState], merge_debate_state]
    risk_debate_state: Annotated[Optional[RiskDebateState], merge_risk_debate_state]
    
    # Research debate state for multi-round bull/bear research debates
    research_debate_state: Annotated[Optional[dict], update_value]
    
    # Investment and trading plans
    investment_plan: Annotated[Optional[str], update_value]
    trader_investment_plan: Annotated[Optional[str], update_value]
    final_trade_decision: Annotated[Optional[str], update_value]
    
    # TASK 7.5.2: Add aggregation ready flag and report readiness tracking
    aggregation_ready: Annotated[Optional[bool], update_value]
    report_readiness: Annotated[Optional[dict], update_value]
